### PR TITLE
openpgp: another rewrite of UID/self-sig parsing

### PR DESCRIPTION
**Pull request by maxtaco**
- In this rewrite, we're not making a subloop for consuming signatures after reading a UID;
  we're just dropping into the loop that's already in place. We'll look for a self-sig
  so long as there's not a valid self-sig already consumed. We only take the UserID as valid
  once we find a valid self-signature. Otherwise, we'll just silently drop that UserID.
- This should fix CORE-2344, and there's a test case for it that now works.
- Also add another test case for Keys without any valid UIDs.

### CODE REVIEW

**Comment by maxtaco** [comment]
@oconnor663 can you take a look at this? We should discuss it in person, since it's subtle, and I'm not 100% convinced it's correct.

**Comment by maxtaco** [comment]
(This is a fix for keybase/client#1685)

**Comment by oconnor663** [comment]
In my entirely unqualified opinion, this change looks good.

File: openpgp/keys.go [hunk 1]
```diff
@@ -353,33 +353,22 @@ EachPacket:
 			current = new(Identity)
```
Comment (oconnor663) [inline openpgp/keys.go:357]: While we're at it, could we comment the shit out of all of these cases? :-D

**Comment by aviau** [comment]
@maxtaco Do I have your permission to upstream this fix?

**Comment by maxtaco** [comment]
Yes!

**Pull request merged**